### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by dangduomg**
# Changes
## 1. Use CMake build system
Makefile is a pain in the ass, especially for multi-platform building, not to mention that LLVM supports it, so I want to use CMake instead.
## 2. Refactor the lexer
The lexer, previously spread out in 3 files, is now consolidated into one file according to the "one header, one class, one source" principle.
Some parts of code are refactored. For example, the `if` chain in the lexer's main loop is replaced with more compact `switch` statement.

### CODE REVIEW

**Comment by dangduomg** [comment]
don't close it prematurely @SohamTilekar!!!!!
give me feedback, have some discussions

**Comment by SohamTilekar** [comment]
final makefile should contain the cppcheck, building with cl, and gcc differently pass the Paramar to the exe ./test/testscipt.gc as default by the post code runner.

**Comment by dangduomg** [comment]
it succeeded in building with `cl`
see the workflow

**Comment by SohamTilekar** [comment]
Bro I Accepted the K&R style.

**Comment by dangduomg** [comment]
yayy

File: src/main.cpp [hunk 1]
```diff
@@ -2,10 +2,10 @@
 #include <fstream>
 #include <string>
 #include "lexer/lexer.hpp"
-#include "parser/parser.hpp"
+// #include "parser/parser.hpp"
 
 #define DEBUG_LEXER true
```
Comment (SohamTilekar) [inline src/main.cpp:7]: make it false before pushing code.

Comment (SohamTilekar) [reply src/main.cpp:7]: TO making it automatically false you can setup the git hook,
by Putting the following code in the .git/hooks/pre-commit.
image

```shell script
#!/bin/sh

# List all staged files
staged_files=$(git diff-index --cached --name-only HEAD)

# Check if a specific file is staged
FILE_PATH="src/main.cpp"
if echo "$staged_files" | grep -q "$FILE_PATH"; then
    if git diff --cached --name-only --diff-filter=ACM | grep --quiet "$FILE_PATH"; then
        if grep -q '#define DEBUG_LEXER true' "$FILE_PATH"; then
            sed -i 's/#define DEBUG_LEXER true/#define DEBUG_LEXER false/g' "$FILE_PATH"
        fi
        if grep -q '#define DEBUG_PARSER true' "$FILE_PATH"; then
            sed -i 's/#define DEBUG_PARSER true/#define DEBUG_PARSER false/g' "$FILE_PATH"
        fi
        git add "$FILE_PATH"
    fi
fi

# Exit with non-zero status to abort the commit if needed
exit 0
```

**Pull request merged**